File: Ass_4_8_1.java
import java.util.HashMap;
import java.util.Map;

// Helper Class: Represents a node in the Trie
class ContactTrieNode {
    Map<Character, ContactTrieNode> children;
    boolean isEndOfWord;

    public ContactTrieNode() {
        children = new HashMap<>();
        isEndOfWord = false;
    }
}

// Helper Class: The Trie data structure
class ContactTrie {
    private final ContactTrieNode root;

    public ContactTrie() {
        root = new ContactTrieNode();
    }

    // Insert a contact name
    public void insert(String word) {
        ContactTrieNode current = root;
        for (char ch : word.toCharArray()) {
            current.children.putIfAbsent(ch, new ContactTrieNode());
            current = current.children.get(ch);
        }
        current.isEndOfWord = true;
    }

    // Search for a complete contact name
    public boolean search(String word) {
        ContactTrieNode current = root;
        for (char ch : word.toCharArray()) {
            if (!current.children.containsKey(ch)) {
                return false;
            }
            current = current.children.get(ch);
        }
        return current.isEndOfWord;
    }

    // Check if any contact starts with a given prefix
    public boolean startsWith(String prefix) {
        ContactTrieNode current = root;
        for (char ch : prefix.toCharArray()) {
            if (!current.children.containsKey(ch)) {
                return false;
            }
            current = current.children.get(ch);
        }
        return true;
    }
}

// Main Class (Must match filename Ass_4_8_1.java)
public class Ass_4_8_1 {
    public static void main(String[] args) {
        ContactTrie contactTrie = new ContactTrie();

        // 1. Input Contacts
        String[] contacts = {"Anil", "Anitha", "Anirudh", "Bala", "Balaji"};
        System.out.println("--- Inserting Contacts ---");
        for (String name : contacts) {
            contactTrie.insert(name);
        }

        System.out.println("\n--- Performing Operations ---");

        // Operation 1: Search for contact "Anil"
        boolean result1 = contactTrie.search("Anil");
        System.out.println("Contact 'Anil' found → " + result1);

        // Operation 2: Search for contact "Anand"
        boolean result2 = contactTrie.search("Anand");
        System.out.println("Contact 'Anand' not found → " + result2);

        // Operation 3: Check prefix "Ani"
        boolean result3 = contactTrie.startsWith("Ani");
        System.out.println("Prefix 'Ani' exists → " + result3);

        // Operation 4: Check prefix "Bal"
        boolean result4 = contactTrie.startsWith("Bal");
        System.out.println("Prefix 'Bal' exists → " + result4);
    }
}

File: Ass_4_8_1.py
class TrieNode:
    """
    Represents a single node in the Trie.
    """
    def __init__(self):
        # Dictionary to store children nodes (Key: Char, Value: TrieNode)
        self.children = {}
        # Boolean to mark if this node represents the end of a complete name
        self.is_end_of_word = False

class Trie:
    """
    The main Trie class to manage contact operations.
    """
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        """
        Inserts a contact name into the Trie.
        """
        current = self.root
        for char in word:
            # If the character doesn't exist, create a new node
            if char not in current.children:
                current.children[char] = TrieNode()
            # Move to the next node
            current = current.children[char]
        
        # Mark the end of the word
        current.is_end_of_word = True

    def search(self, word):
        """
        Checks if a full contact name exists.
        Returns: True if found, False otherwise.
        """
        current = self.root
        for char in word:
            if char not in current.children:
                return False
            current = current.children[char]
        
        # It must be the end of a stored word, not just a prefix
        return current.is_end_of_word

    def starts_with(self, prefix):
        """
        Checks if any contact starts with the given prefix.
        Returns: True if exists, False otherwise.
        """
        current = self.root
        for char in prefix:
            if char not in current.children:
                return False
            current = current.children[char]
        
        # If we successfully traversed the prefix, return True
        return True

# --- Driver Code (based on your Test Case) ---

def run_mobile_contact_test():
    # 1. Initialize Trie
    contact_trie = Trie()
    
    # 2. Input Contacts
    contacts = ["Anil", "Anitha", "Anirudh", "Bala", "Balaji"]
    print(f"--- Inserting Contacts: {contacts} ---")
    for name in contacts:
        contact_trie.insert(name)
    
    print("\n--- Performing Operations ---")

    # Operation 1: Search for contact "Anil"
    result_1 = contact_trie.search("Anil")
    print(f"Contact 'Anil' found → {result_1}")

    # Operation 2: Search for contact "Anand"
    result_2 = contact_trie.search("Anand")
    print(f"Contact 'Anand' not found → {result_2}")

    # Operation 3: Check prefix "Ani"
    result_3 = contact_trie.starts_with("Ani")
    print(f"Prefix 'Ani' exists → {result_3}")

    # Operation 4: Check prefix "Bal"
    result_4 = contact_trie.starts_with("Bal")
    print(f"Prefix 'Bal' exists → {result_4}")

# Execute the test
if __name__ == "__main__":
    run_mobile_contact_test()